### PR TITLE
S2U-27 Messages: Convert a private message into a forum message in the FAQ forum

### PR DESCRIPTION
**Pull request by stetsche**
https://sakaiproject.atlassian.net/browse/SAK-35033

Conversion scripts: https://github.com/sakaiproject/sakai-reference/pull/204

### CODE REVIEW

File: msgcntr/messageforums-app/pom.xml [hunk 1]
```diff
@@ -88,6 +88,10 @@
 			<groupId>org.springframework</groupId>
 			<artifactId>spring-core</artifactId>
 		</dependency>
+        <dependency>
```
Comment (jesusmmp) [inline msgcntr/messageforums-app/pom.xml:91]: Spacing

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java [hunk 2]
```diff
@@ -8138,6 +8153,10 @@ public String getEditorRows() {
 		return ServerConfigurationService.getString("msgcntr.editor.rows", "22");
 	}
 
+	public String getEditorCols() {
+		return "132";
```
Comment (jesusmmp) [inline msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:8157]: We should avoid magic numbers ;)

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AreaManagerImpl.java [hunk 3]
```diff
@@ -48,6 +48,7 @@
 import org.sakaiproject.tool.api.Placement;
 import org.sakaiproject.tool.api.SessionManager;
 import org.sakaiproject.tool.api.ToolManager;
+import org.sakaiproject.user.api.UserDirectoryService;
```
Comment (jesusmmp) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AreaManagerImpl.java:51]: Undo this change and line 223. No changes on this file

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java [hunk 4]
```diff
@@ -710,6 +733,58 @@ public ActorPermissions createDefaultActorPermissions()
        return actorPermissions;
     }
     
+    // Create a new FAQ forum based on defaults
+    private DiscussionForum createFaqForum(Area discussionArea) {
+        log.debug("Creating a new FAQ Forum");
+
+        // Locale locale = Locale.getDefault();
```
Comment (jesusmmp) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java:740]: Remove comment

File: msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/Topic.hbm.xml [hunk 5]
```diff
@@ -191,6 +191,9 @@
 	    <property name="restrictPermissionsForGroups">
 	      <column name="RESTRICT_PERMS_FOR_GROUPS" not-null="true" default="0" />
 	    </property>
+      <property name="faqTopic">
```
Comment (jesusmmp) [inline msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/Topic.hbm.xml:194]: Spacing

Comment (mpellicer) [reply msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/Topic.hbm.xml:194]: Spacing vs MSGCNTR, name a better DUO!

**Comment by mpellicer** [comment]
Thanks for helping me with the reviews @jesusmmp !

**Comment by stetsche** [comment]
@jesusmmp @mpellicer Thanks for taking a look! I addressed the issues together with a few other minor changes﻿.
I also added the translations that @NuriaQS and @DavidVegaSa kindly provided.
I think the PR is ready now!

**Comment by stetsche** [comment]
Hi @PabloSanRoman﻿, could you please help with the following translations?

```
pvt_publish_to_faq=Publish to FAQ
pvt_title=Title
pvt_question=Question
pvt_question_title_prefix=Q:
pvt_answer=Answer
pvt_answer_title_prefix=A:
pvt_answer_info=You can leave the answer field empty if you don\u0027t want to post it yet.
pvt_missing_title=You must enter a title before you can post to FAQ.
pvt_missing_question=You must enter a question before you can post to FAQ.

cdfm_faq_forum_title=FAQ Forum
cdfm_faq_forum_description=Forum for frequently asked questions
cdfm_faq_topic_title=FAQ Topic
cdfm_faq_topic_description=Topic for frequently asked questions
```

Greetings,
Markus

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java [hunk 6]
```diff
@@ -4285,11 +4287,50 @@ public String processDfMsgReplyMsgFromEntire()
...
+    String answerPrefix = getResourceBundleString(MSG_PVT_ANSWER_PREFIX);
+    // String targetPrefix = Boolean.TRUE.equals(selectedTopic.getTopic().getFaqTopic())
+    //     ? getResourceBundleString(MSG_PVT_ANSWER_PREFIX)
+    //     : replyPrefix;
+    String questionPrefix = getResourceBundleString(MSG_PVT_QUESTION_PREFIX);
+    String title = StringUtils.trim(selectedMessage.getMessage().getTitle());
+
+    if (StringUtils.startsWith(title, replyPrefix)) {
+        // Re: title --> Re: title
+        log.debug("LOL 1.0");
```
Comment (jesusmmp) [inline msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:4304]: Remove this one and the others LOL xD

File: msgcntr/messageforums-app/src/webapp/jsp/privateMsg/pvtMsgPublishToFaqEdit.jsp [hunk 7]
```diff
@@ -0,0 +1,90 @@
+<%-- TODO Master merge:
```
Comment (jesusmmp) [inline msgcntr/messageforums-app/src/webapp/jsp/privateMsg/pvtMsgPublishToFaqEdit.jsp:1]: Great annotations!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/evaluation/QuestionScoresBean.java [hunk 8]
```diff
@@ -182,6 +182,9 @@ public class QuestionScoresBean implements Serializable, PhaseAware {
   @Setter @Getter
   private boolean emiItemPresent;
 
+  @Setter @Getter
+  private boolean randomItemPresent;
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/evaluation/QuestionScoresBean.java:186]: This code is not related to this patch.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/util/ItemCancellationUtil.java [hunk 9]
```diff
@@ -17,46 +17,38 @@
 package org.sakaiproject.tool.assessment.util;
 
 import org.sakaiproject.tool.assessment.data.ifc.assessment.ItemDataIfc;
+import org.sakaiproject.tool.assessment.data.ifc.assessment.SectionDataIfc;
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/util/ItemCancellationUtil.java:20]: Review all the samigo code. Not related to this feature.

**Comment by stetsche** [comment]
Thanks @jesusmmp﻿, done!

**Review by jesusmmp — APPROVED** [review]

File: msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/Topic.hbm.xml (hunk 5 above)
Comment (stetsche) [reply msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/Topic.hbm.xml]: Yes, I love it when i have to mix tabs with spaces to "fix" the indentation xD

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java (hunk 6 above)
Comment (stetsche) [reply msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java]: :sweat_smile:

**Comment by stetsche** [comment]
Codacy issues seem like all false-positives to me.

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties [hunk 10]
```diff
@@ -891,3 +891,19 @@ perm-msg.permissions.allowToField.users=Pertsona guztiei bidali
 perm-msg.permissions.allowToField.myGroups=Norberaren taldeei bidali
 perm-msg.permissions.allowToField.myGroupMembers=Norberaren taldeen pertsonei bidali
 perm-msg.permissions.allowToField.myGroupRoles=Norberaren taldeen rolei bidali
+
+#S2U-27
+pvt_publish_to_faq=Publish to FAQ
```
Comment (mpellicer) [inline msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:896]: Hey @pablosanroman, we have a small number of new keys, would you mind translating them? Not in a hurry.

**Comment by mpellicer** [comment]
This looks great, good job monster @stetsche ! I'm missing the conversion scripts, would you mind contributing and linking them?

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties (hunk 10 above)
Comment (PabloSanRoman) [reply msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:896]: Hi @mpellicer!here it is!: 

pvt_title=Titulua
pvt_question=Galdera
pvt_question_title_prefix=Q:
pvt_answer=Erantzuna
pvt_answer_title_prefix=A:
pvt_answer_info=Erantzunaren eremua hutsik utz dezakezu publikatu nahi ez baduzu oraindik. 
pvt_missing_title=Ezinbestekoa da izenburu bat sartzea "Ohiko galderak" atalean publikatzeko. 
pvt_missing_question=Ezinbestekoa da galdera bat sartzea "Ohiko galderak" atalean publikatzeko. 

cdfm_faq_forum_title=FAQ foroa
cdfm_faq_forum_description=Foroa Ohiko galderentzat
cdfm_faq_topic_title=FAQ Gaia
cdfm_faq_topic_description=Gaia Ohiko galderentzat
